src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,282 @@
+# -----------------------------------------------------------------------------
+# Author: A. B. G. Chalk, STFC Daresbury Lab
+
+'''This module provides the sclarization transformation class.'''
+
+import itertools
+from typing import Optional, Dict, Any
+
+from psyclone.core import VariablesAccessInfo, Signature
+from psyclone.psyGen import Kern
+from psyclone.psyir.nodes import Call, CodeBlock,  \
+        Loop, Reference, Routine
+from psyclone.psyir.symbols import DataSymbol
+from psyclone.psyir.transformations.loop_trans import LoopTrans
+
+
+class ScalarizationTrans(LoopTrans):
+    '''This transformation takes a Loop and converts any array accesses
+    to scalar if the results of the loop are unused, and the initial value
+    is unused. For example:
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.transformations import ScalarizationTrans
+    >>> from psyclone.psyir.nodes import Loop
+    >>> code = """program test
+    ... integer :: i,j
+    ... real :: a(100), b(100)
+    ... do i = 1,100
+    ...   a(i) = i
+    ...   b(i) = a(i) * a(i)
+    ... end do
+    ... do j = 1, 100
+    ...  if(b(i) > 200) then
+    ...    print *, b(i)
+    ...  end if
+    ... end do
+    ... end program"""
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> scalarise = ScalarizationTrans()
+    >>> scalarise.apply(psyir.walk(Loop)[0])
+    >>> print(FortranWriter()(psyir))
+    program test
+      integer :: i
+      integer :: j
+      real, dimension(100) :: a
+      real, dimension(100) :: b
+      real :: a_scalar
+    <BLANKLINE>
+      do i = 1, 100, 1
+        a_scalar = i
+        b(i) = a_scalar * a_scalar
+      enddo
+      do j = 1, 100, 1
+        if (b(i) > 200) then
+          ! PSyclone CodeBlock (unsupported code) reason:
+          !  - Unsupported statement: Print_Stmt
+          PRINT *, b(i)
+        end if
+      enddo
+    <BLANKLINE>
+    end program test
+    <BLANKLINE>
+    '''
+
+    @staticmethod
+    def _is_local_array(signature: Signature,
+                        var_accesses: VariablesAccessInfo) -> bool:
+        '''
+        :param signature: The signature to check if it is a local array symbol
+                          or not.
+        :param var_accesses: The VariableAccessesInfo object containing
+                             signature.
+        :returns: whether the symbol corresponding to signature is a
+                  local array symbol or not.
+        '''
+        if not var_accesses[signature].is_array():
+            return False
+        base_symbol = var_accesses[signature].all_accesses[0].node.symbol
+        if not base_symbol.is_automatic:
+            return False
+
+        return True
+
+    @staticmethod
+    def _have_same_unmodified_index(signature: Signature,
+                                    var_accesses: VariablesAccessInfo) \
+            -> bool:
```

Comment:
This is pedantic but since I still have an open comment above, we try to avoid backslashes, could you do:
```
    def _have_same_unmodified_index(
            signature: Signature,
            var_accesses: VariablesAccessInfo
    ) -> bool:
```
And similar in the backslashes in signatures below.